Numeric testcase for precision overflow triggers known protocol violation

bsatarupa opened the pull request:
### Description
Before Fix:
Numeric testcase for precision overflow triggers known protocol violation

After Fix:
Postgres can store numeric data upto 1000 digits but SQL server has a maximum limit of 38 digits. Since, adding the maximum numeric precision check on TDS sender side results into protocol violation, have addressed the issue on engine side by calculating exact numeric digits from Postgres numeric-blocks format and reporting error whenever #digits > TDS_MAX_NUMERIC_PRECISION(i.e. 38).

Task: BABEL-3179
Signed-off-by: Satarupa Biswas <satarupb@amazon.com>


- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

surendravishnoi commented on src/backend/utils/adt/numeric.c:
```diff
@@ -5963,6 +5964,14 @@ numeric_avg(PG_FUNCTION_ARGS)
 	init_var(&sumX_var);
 	accum_sum_final(&state->sumX, &sumX_var);
+
+	if (sql_dialect == SQL_DIALECT_TSQL &&
```
let us move dialect check inside hook

bsatarupa replied:
done

surendravishnoi commented on src/backend/utils/adt/numeric.c:
```diff
@@ -5995,6 +6004,14 @@ numeric_sum(PG_FUNCTION_ARGS)
 	accum_sum_final(&state->sumX, &sumX_var);
+
+	if (sql_dialect == SQL_DIALECT_TSQL &&
+	    (*exact_numeric_digit_count_hook)(sumX_var.weight, sumX_var.dscale, sumX_var.digits[0], DEC_DIGITS) >
```
we should abstract the overflow detection inside hook, say detect_numeric_overflow which returns true or false. That way TSQLMaxNumPrecision is not used on engine side

bsatarupa replied:
done

Commit:
Numeric testcase for precision overflow triggers known protocol violation

Fix:
Postgres can store numeric data upto 1000 digits but SQL server has a maximum limit of 38 digits. Since, adding the maximum numeric precision check on TDS sender side results into protocol violation, have addressed the issue on engine side by calculating exact numeric digits from Postgres numeric-blocks format and reporting error whenever #digits > TDS_MAX_NUMERIC_PRECISION(i.e. 38).

Task: BABEL-3179
Signed-off-by: Satarupa Biswas <satarupb@amazon.com>
